book/src/proptest/state-machine.md

```diff
@@ -0,0 +1,174 @@
+  ```
+  
+  If your `ReferenceStateMachine::init_state` uses a non-constant strategy, you have to use the `ref_state` to initialize this to a corresponding state to ensure that you have consistent initial states.
+
+- Apply the `transition` on the SUT state:
+  
+  ```rust,ignore
+  fn apply(
+    mut state: Self::SystemUnderTest,
+    ref_state: &Self::Reference::State,
+    transition: Transition
+  ) -> Self::SystemUnderTest
+  ```
+  
+  This is also where you'll want to check any post-conditions that apply to a given transition, so after you apply the transition to the state, you can `assert!` some properties. Alternatively or additionally, you can use the `ref_state` for comparison, which will have the same transition that is given to this function already applied to it.
+
+- Check properties that apply in any state:
+
+  ```rust,ignore
+  fn check_invariants(state: &Self::SystemUnderTest, ref_state: &Self::Reference::State)
+  ```
+
+  These must always hold and will be checked after every transition. Just like with `apply`, you have the option to use the `ref_state` for comparison.
+
+To add some teardown logic to run at the end of each test case, you can override the `teardown` function, which by default simply drops the state:
+
+```rust,ignore
+fn teardown(state: Self::SystemUnderTest)
+```
+
+### Make the state machine test runnable
+
+Finally, to run the `StateMachineTest`, you can use the `prop_state_machine!` macro. For example:
+
+```rust,ignore
+prop_state_machine! {
+  #[test]
+  fn name_of_the_test(sequential 1..20 => MyStateMachineTest);
+}
+```
+
+You pick a `name_of_the_test` and a single numerical value or a range after the `sequential` keyword for a number of transitions to be generated for the state machine execution. The `MyStateMachineTest` is whatever you've implemented the `StateMachineTest` for.
+
+And that's it. You can run the test, perhaps with `cargo watch` as you develop it further, and see if it can find some interesting counter-examples to your properties.
+
+### Extra tips
+
+Because a state machine test may be heavier than regular prop tests, if you're running your tests in a CI you may want to override the default `proptest_config`'s `cases` to include fewer cases in a single run. You can also use `PROPTEST_CASES` environment variable and during development it is preferable to override this to run many cases to get a better chance of catching those pesky ~~bugs~~, erm, defects.
+
+> Given that there are thought to be in the region of another four million species that we have not yet even named, there is no doubt that scientists will be kept happily occupied studying them for millennia, so long as the insects remain to be studied. Would the world not be less rich, less surprising, less wonderful, if these peculiar creatures did not exist?
+>
+> -- <cite>Dave Goulson, Silent Earth</cite>
+
+So let's leave bugs alone and only squash defects instead!
```

Comment:
I know this is trying to be a bit fun and i'm being somewhat of a downer, but i think this interlude about bugs actually detracts from the clarity of the documentation. Everything else I've found to be well written and on topic without being too terse nor too wordy, and then this kind of left my head scratching and I had to re-read it. If you don't want to use the word "bug", that's totally fine, we can use the word "defect" without the need to elaborate on that. Keep in mind too this is lib documentation, not a blog, which people want to scan quickly (meaning less distraction is better) while still getting enough information (and related reference links) they need to do the job.

Reply:
> if you're running your tests in a CI you may want to override the default `proptest_config`'s `cases` to include fewer cases in a single run

I'm not sure what's most common, but in my case CI is has significantly more horsepower than dev's laptops. Usually we'll keep dev test runs small (multiples of 10) so they can iterate quickly, and CI large (multiples of 1000). So I'm not sure if this is the most common recommendation. Maybe it would be better to just acknowledge that there is a parameter to configure number of test cases and that it's good practice to configure this parameter separately between dev and CI environments based on the organizations needs?